Modules.md

```diff
@@ -170,30 +231,154 @@ reserve up to the *maximum* is not. When a maximum memory size is *not* declared
+which constitute most of the byte size of the module, near the end so that all
+metadata necessary for recursive module loading and parallel compilation is
+available before compilation begins.
+
+## Definition Index Space
+
+The *definition index space* represents the union of all definitions introduced
+by the [import](#imports), [global](#global-section), [memory](#linear-memory-section),
+[table](#table-section), and [code](#code-section) sections. Each of these
+sections can introduce zero or more definitions and the definition index space
+simply assigns monotonically increasing indices to these definitions according
+to their absolute order as defined in [BinaryEncoding.md](BinaryEncoding.md).
+
+The definition index space is used by:
+* [exports](#exports), to indicate which definition to export
+* [elements sections](#elements-section), to place definitions into tables
+
+In the future, an `address_of` operator could be added which, given a definition
+index immediate, returns a first-class [reference](GC.md) to the definition.
+
+Note: the definition index spaces is a validation/compile-time concept, not
+runtime state of an instance.
+
+## Function Index Space
+
+The *function index space* represents the subsequence of the 
+[definition index space](#definition-index-space) obtained by discarding
+non-function definitions.
+
+The function index space is used by:
+* [calls](AstSemantics.md#calls), to identify the callee of a direct call
+
+## Global Index Space
+
+The *global index space* represents the subsequence of the
+[definition index space](#definition-index-space) obtained by discarding
+non-global definitions.
+
+The global index space is used by:
+* [global variable access operators](AstSemantics.md#global-variables), to
+  identify the global variable to read/write
+* [data segments](#data-section), to define the offset of a data segment
+  (in linear memory) as the value of a global variable
+
+## Linear Memory Index Space
+
+The *linear memory index space* represents the subsequence of the
+[definition index space](#definition-index-space) obtained by discarding
+non-linear-memory definitions.
+
+The linear memory index space is only used by the 
+[data section](#data-section). In the MVP, there is at most one linear memory so
+this index space is just a placeholder for when there can be 
+[multiple memories](FutureFeatures.md#multiple-tables-and-memories).
+
+## Table Index Space
+
+The *table index space* represents the subsequence of the
+[definition index space](#definition-index-space) obtained by discarding
+non-table definitions.
+
+The table index space is only used by the [elements section](#elements-section).
+In the MVP, there is at most one table so this index space is just
+a placeholder for when there can be 
+[multiple tables](FutureFeatures.md#multiple-tables-and-memories).
+
+## Initializer Expression
+
+Initializer expressions are evaluated at instantiation time and are currently
```

Comment:
By the time the start function has run, the data segments must have already been copied into linear memory; the key is being able to, one way or another, use an instantiation-time-provided value to determine the offset.  For mutable global variables of course we could do zero-initialization, but for *immutable* global variables we need *some* initializer.  We could simply use a type-appropriate numeric immediate (and then we'll have to figure out something for globals with GC reference type), but it felt slightly more regular to use this initializer-expression concept if it was already present for data segment offsets since this affords us greater flexibility in the future.